### PR TITLE
drag drop textboxes to dashboard and choose the boxes background color from a palette

### PR DESCRIPTION
**Pull request by anhthu253**
The dashboard page contains:
- [ ] some dummy textboxes
- [ ] one or several boards
- [ ] a color palette popped up by right clicking on any textbox
- [ ] choice of colors from the palette for the textbox
- [ ] drag and drop textboxes in the boards
- [ ] add and delete boards

### CODE REVIEW

File: pages/dashboard.js [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+
+  useEffect(() => {
+    //close the color palette by clicking outside it
+    document.body.addEventListener("click", (event) => {
+      if (event.target.id !== "ColorPalette")
+        setColorPalette((pl) => {
+          return { ...pl, display: false };
+        });
+    });
+  }, []);
```
Comment (Kheirah) [inline pages/dashboard.js]: You are adding an event listener to the body, but you're never removing it. As a result, the onClick event is added every time you visit the dashbard page, which will cause unexpected behavior.

File: pages/dashboard.js [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+        ondrop={(event) => {
+          drop(event);
+          setRemainDragItemCount((prevCount) => prevCount + 1);
+        }}
+        onRightClick={(event, id) => openColorPalette(event, id)}
+      ></DragContainer>
+
+      <DropzoneContainer>
+        {dropzones.map((dropzone) => (
+          <DZWrapper key={Math.random().toString(36).substring(2)}>
```
Comment (Kheirah) [inline pages/dashboard.js]: The key prop needs a unique, yet stable value. You are providing a unique value, but it's not stable across render cycles. A random key value results in rendering all items anew due to the key prop changing every time the component is re-rendered. You always need a stable, unique key value.

**Review by Kheirah — CHANGES_REQUESTED** [review]
Overall very good! Just have a look at the places where you need to change or add something =)

File: pages/dashboard.js (hunk 1 above)
Comment (anhthu253) [reply pages/dashboard.js]: Hi Ernst, thank you very much for the review! You meant I should remove the event when the click happens inside the palette?

File: pages/dashboard.js (hunk 2 above)
Comment (anhthu253) [reply pages/dashboard.js]: aaah thank you for telling me this. I will change it :)

File: pages/dashboard.js (hunk 1 above)
Comment (Kheirah) [reply pages/dashboard.js]: No, I mean remove the listener when the component is unmounted. There's a built-in way to do that within useEffect 😄

Comment (anhthu253) [reply pages/dashboard.js]: ah ok i understand :D

**Comment by anhthu253** [comment]
Btw it s not related to this PR but can u pls recommend me an event which should be triggered before leaving a page, something similar to beforeunload event?

File: pages/dashboard.js [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+        });
+    });
+    return () => {
+      document.body.removeEventListener("click", (event) => {
+        if (event.target.id !== "ColorPalette")
+          setColorPalette((pl) => {
+            return { ...pl, display: false };
+          });
+      });
+    };
```
Comment (Kheirah) [inline pages/dashboard.js]: Very good!

If you wanted to slim down the code further you could write a function for the callback and hand down this function in addEventListener and removeEventListener. This is optional, you choose :)

**Review by Kheirah — APPROVED** [review]
Looking good!

**Comment by Kheirah** [comment]
> Btw it s not related to this PR but can u pls recommend me an event which should be triggered before leaving a page, something similar to beforeunload event?

You could use the callback in the return of the useEffect hook to do that.

File: pages/dashboard.js (hunk 3 above)
Comment (anhthu253) [reply pages/dashboard.js]: i know i was so lazy. I m having many problems with the other PR it drove me quite mad 🙈 😅. But I will do it!

**Comment by anhthu253** [comment]
> > Btw it s not related to this PR but can u pls recommend me an event which should be triggered before leaving a page, something similar to beforeunload event?
> 
> You could use the callback in the return of the useEffect hook to do that.

yes I could do that. Thank you!

**Pull request merged**